appinventor/appengine/src/com/google/appinventor/client/wizards/TemplateUploadWizard.java

```diff
@@ -349,27 +349,70 @@ public void execute() {
     initFinishCommand(new Command() {
         @Override
         public void execute() {
-          String filename = selectedTemplateNAME + PROJECT_ARCHIVE_EXTENSION;
           // Make sure the project name is legal and unique.
-          if (!TextValidators.checkNewProjectName(selectedTemplateNAME)) {
+          if (!TextValidators.checkNewProjectName(selectedTemplateNAME, true)) {
             center();
-            return;
-          }
-          NewProjectCommand callbackCommand = new NewProjectCommand() {
+            
+            String suggestedName = "";
+            String title;
+            selectedTemplateNAME = selectedTemplateNAME.replace(" ", "_")
+                .replaceAll("[^a-zA-Z0-9_]", "");
+
+            if (TextValidators.isTitleReserved()) {
+              title = MESSAGES.reservedTitleFormatError() + " : " + selectedTemplateNAME;
+            } else if (TextValidators.isTitleInvalid()) {
+              suggestedName = ProjectUploadWizard.getSuggestedName(selectedTemplateNAME);
+              title = MESSAGES.invalidTitleFormatError();
+              if (!TextValidators.checkNewProjectName(suggestedName, true)) {
+                  suggestedName = "";
+                  title += " : " + selectedTemplateNAME;
+              } else {
+                title += MESSAGES.suggestNameTitleCaption();
+              }
+            } else {
+              suggestedName = ProjectUploadWizard.getSuggestedName(selectedTemplateNAME);
+              title = MESSAGES.duplicateTitleFormatError();
+              if (!TextValidators.checkNewProjectName(suggestedName, true)) {
+                  suggestedName = "";
+                  title += " : " + selectedTemplateNAME;
+              } else {
+                title += MESSAGES.suggestNameTitleCaption();
+              }
+            }
```

Comment:
Given that a lot of this code seems a duplicate of ProjectUploadWizard you may want to refactor the logic into one of the other utility classes and just call that from here.